colossalai/gemini/chunk/chunk.py

```diff
@@ -306,46 +398,56 @@ def copy_tensor_to_chunk_slice(self, tensor: torch.Tensor, data_slice: torch.Ten
         self.chunk_total[tensor_info.offset:tensor_info.end].copy_(data_slice.data.flatten())
         tensor.data = self.chunk_total[tensor_info.offset:tensor_info.end].view(tensor.shape)
 
-    @property
-    def can_move(self) -> bool:
-        return not self.is_gathered
-
-    @property
-    def can_release(self) -> bool:
+    def get_valid_length(self) -> int:
+        """Get the valid length of the chunk's payload.
+        """
         if self.keep_gathered:
-            return False
+            return self.utilized_size
         else:
-            return self.tensors_state_monitor[TensorState.HOLD] + \
-                   self.tensors_state_monitor[TensorState.HOLD_AFTER_BWD] == self.num_tensors
+            return self.valid_end
 
-    @property
-    def can_reduce(self):
-        return self.tensors_state_monitor[TensorState.READY_FOR_REDUCE] == self.num_tensors
-
-    @property
-    def has_inf_or_nan(self) -> bool:
+    def init_pair(self, friend_chunk: 'Chunk') -> None:
+        """Initialize the paired chunk.
         """
-        Check if the chunk has inf or nan values in CUDA.
+        if self.paired_chunk is None and friend_chunk.paired_chunk is None:
+            self.paired_chunk = friend_chunk
+            friend_chunk.paired_chunk = self
+        else:
+            assert self.paired_chunk is friend_chunk
+            assert friend_chunk.paired_chunk is self
+
+    def optim_update(self) -> None:
+        """Update the fp16 chunks via their fp32 chunks. It's used by the optimizer.
         """
-        if self.is_gathered:
-            valid_tensor = self.chunk_total[:self.utilized_size]
+        # sanity check
+        assert self.paired_chunk is not None
+
+        friend_chunk = self.paired_chunk
+        if self.is_gathered is True:
+            assert friend_chunk.is_gathered is True
+            self.chunk_total.copy_(friend_chunk.chunk_total)
+            self.optim_sync_flag = True
+        elif friend_chunk.device_type == 'cuda' and self.device_type == 'cuda':
+            self.cuda_shard.copy_(friend_chunk.cuda_shard)
+            self.optim_sync_flag = True
+            self.cpu_vis_flag = False
         else:
-            assert self.cuda_shard is not None    # only check in CUDA
-            valid_tensor = self.cuda_shard[:self.valid_end]
+            assert friend_chunk.device_type == 'cpu'

```

Comment:
Please add some comments to explain where the fp16 chunk will be updated when they are on CPU.